bibliopixel/drivers/spi_driver_base.py

```diff
@@ -1,77 +1,168 @@
 
-        # File based SPI requires a bytearray so we have to overwrite _buf
-        if self.use_py_spi:
-            a, b = -1, -1
-            d = self.dev.replace('/dev/spidev', '')
-            s = d.split('.')
-            if len(s) == 2:
-                a = int(s[0])
-                b = int(s[1])
+    def send_packet(self, data):
+        raise NotImplementedError
 
-            if a < 0 or b < 0:
-                error(BAD_FORMAT_ERROR)
+    def compute_packet(self, data):
+        return data
 
-            self._bootstrapSPIDev()
-            import spidev
-            self.spi = spidev.SpiDev()
-            self.spi.open(a, b)
-            self.spi.max_speed_hz = int(self._spiSpeed * 1000000.0)
-            log.info('py-spidev speed @ %.1f MHz',
-                     (float(self.spi.max_speed_hz) / 1000000.0))
-        else:
-            self._buf = bytearray(self._buf)
-            self.spi = self._open(self.dev, 'wb')
-
-    def _bootstrapSPIDev(self):
-        if self.use_py_spi:
-            try:
-                import spidev
-            except:
-                error(CANT_IMPORT_SPIDEV_ERROR)
-
-        if not os.path.exists(self.dev):
+
+# partial source @see https://armbedded.taskit.de/node/318
+# @see <linux/spi/spidev.h>
+# hex numbers from good old strace(1)
+IOC_READ = 0x40000000
+IOC_WRITE = 0x40000000
+
+IOC_BITS_PER_WORD = 0x00016b03
+IOC_MAX_SPEED_HZ = 0x00046b04
+
+
+class SpiFileInterface(SpiBaseInterface):
+    """ using os open/write to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not os.path.exists(self._dev):
             error(CANT_FIND_ERROR)
 
+        try:
+            import fcntl
+            self._fcntl = fcntl
+        except ImportError:
+            error(CANT_IMPORT_FCNTL_ERROR)
+        self._spi = open(self._dev, 'wb')
+
+        self.speed = self._spi_speed
+        self.bits_per_word = 8
+        log.info('file io spi speed @ %.1f MHz, %d bits per word', self.speed / 1e6, self.bits_per_word)
+
+    def _ioctl(self, command, buffer):
+        self._fcntl.ioctl(self._spi, IOC_WRITE + command, buffer)
+        self._fcntl.ioctl(self._spi, IOC_READ + command, buffer)
+
+    def _set_speed(self, speed):
+        buffer = struct.pack(">I", int(speed * 1e9))  # unint32
+        self._ioctl(IOC_MAX_SPEED_HZ, buffer)
+        self._speed = struct.unpack(">I", buffer)[0] / 1000
+
+    def _get_speed(self):
+        return self._speed
+
+    speed = property(_get_speed, _set_speed)
+
+    def _set_bits_per_word(self, bits):
+        buffer = struct.pack("B", bits)  # uint8
+        self._ioctl(IOC_BITS_PER_WORD, buffer)
+        self._bits_per_word = int.from_bytes(buffer, byteorder='big')
+
+    def _get_bits_per_word(self):
+        return self._bits_per_word
+
+    bits_per_word = property(_get_bits_per_word, _set_bits_per_word)
+
+    def send_packet(self, data):
+        self._set_speed(self._spi_speed)  # set speed again if other program changed it
```

Comment:
I think this line is wrong. :-o

This is the original member variable that's set during the constructor, which never changes after construction.  This means if you set `self.speed` then that setting gets erased the next time you send a packet!

I think what you wanted to say was `self._set_speed(self._speed)` but that's better expressed as `self.speed = self._speed`.

But even that I think is wrong.  :-D

I'm against doing _anything_ every frame "just in case".  In this case, you're doing something that's fairly consumptive, and might even have side-effects (I don't know if changing the speed might cause the interface to reset).

What is this other program that changes the speed behind our back?  

If someone is so hungry for pain that they are running some other program that changes the speed to a wrong value, then _they_ should be setting the value back in _their_ program, and not expect us to do it for them and have everyone else pay a little penalty for their masochism!

Finally, I think you should be getting merging the two variables `self._spi_speed` and `self._speed`.  Just call the result `self._speed` - no need for the redundant `_spi` prefix!